### PR TITLE
Add numerical constrained noise model

### PR DESCRIPTION
**Pull request by SuperN1ck**
This PR adds a very simple numerical constrained noise model by replacing infinite values in the precision diagonal array with a very big number.

### CODE REVIEW

File: jaxfg/noises/_gaussians.py [hunk 1]
```diff
@@ -77,3 +77,33 @@ def whiten_jacobian(
...
+
+
+@register_dataclass_pytree
+@dataclasses.dataclass
+class ConstrainedDiagonalGaussian(DiagonalGaussian):
+    @staticmethod
+    @overrides
+    def make_from_covariance(
+        diagonal: Union[hints.Array, Sequence[float]],
+        penalty: Union[hints.Array, Sequence[float], float, int] = None,
```
Comment (brentyi) [inline jaxfg/noises/_gaussians.py:89]: Perhaps this could just be:
```
penalty: Union[hints.Array, Sequence[float], float] = 1000.0
```

Some typing comments:
- Having both `float` and `int` is also redundant, see [the numeric tower](https://www.python.org/dev/peps/pep-0484/#the-numeric-tower).
- If `None` is a default/possible value, it's nice to be explicit about the type: `Optional[Union[...]]`

**Review by brentyi — COMMENTED** [review]
Thanks Nick!

This is meant to be a port of gtsam's `Constrained` noise model, right? Do you think you could comment on the implementation differences between how infinite precision values are handled here vs in their implementation?

It seems like they have quite a bit of extra logic (for example, in [whiten](https://github.com/devbharat/gtsam/blob/5f15d264ed639cb0add335e2b089086141127fff/gtsam/linear/NoiseModel.cpp#L336)), but it's not immediately obvious why.

With how things are currently set up, I wonder if this logic could just be folded directly into the `DiagonalGaussian` class; it looks like it could be written as a simple upper bound on the precision values.

**Comment by SuperN1ck** [comment]
Thanks for the feedback Brent! Let me answer your questions :)

> This is meant to be a port of gtsam's Constrained noise model, right? Do you think you could comment on the implementation differences between how infinite precision values are handled here vs in their implementation?
> 
> It seems like they have quite a bit of extra logic (for example, in whiten), but it's not immediately obvious why.

Yes it's inspired by GTSAMs implementation. In general the flow of the data can be roughly described by `DistanceMeasure(Whiten(ResidualVector))`, right? When looking at the special case for possible infinite precision values (aka a hard constraint), whitening the residual vector would also lead to infinite precision, resulting in vanishing gradients etc. 

GTSAM handles that problem two fold. First, in their [whitening process](https://github.com/borglab/gtsam/blob/43e8f1e5aeaf11890262722c1e5e04a11dbf9d75/gtsam/linear/NoiseModel.cpp#L374)
they leave the residual `ai` untouched `c(i) = (bi==0.0) ? ai : ai/bi` and therefore do not produce any infinity values. Second, in their [distance measure](https://github.com/borglab/gtsam/blob/43e8f1e5aeaf11890262722c1e5e04a11dbf9d75/gtsam/linear/NoiseModel.cpp#L384) they multiply the untouched residual vector by the penalty `mu_[i]` in `w[i] = v[i] * sqrt(mu_[i]);` (where `v[i]` is `c[i]` in the previous code snippet).

Since in jaxfg we currently do not do a noise specific distance calculation but rather take the global Mahalanobis distance on the whole residual vector (also I currently see no reason to change that), I combined both in `jnp.where`.

> With how things are currently set up, I wonder if this logic could just be folded directly into the DiagonalGaussian class; it looks like it could be written as a simple upper bound on the precision values.

Sounds good to me! If you are fine with not having an explicit class, I would suggest in [`make_from_covariance`](https://github.com/brentyi/jaxfg/blob/b922ad97a83d8e564cbb67273b669500627c7549/jaxfg/noises/_gaussians.py#L51) we could either check for infinite values or even directly for very small elements in `diagonal` and in that case rather than taking the sqrt-root we could assign the penalty directly to the respective position in `sqrt_precision_diagonal`? Another benefit of this is also to prevent users from running into hard to debug errors (e.g. when their covariance is too small).

File: jaxfg/noises/_gaussians.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 from jax import numpy as jnp
 from overrides import overrides
+from jaxlie.utils import get_epsilon
```
Comment (SuperN1ck) [inline jaxfg/noises/_gaussians.py:6]: Ok to use here?

File: tests/test_integration.py [hunk 3]
```diff
@@ -199,3 +199,62 @@ def test_pose_graph_dogleg():
...
+        jaxlie.SE2.from_xy_theta(3.0, 2.0, 1.0),
+        jaxlie.SE2.from_xy_theta(1.0, 2.0, 3.0),
+    ]
+
+    factors: List[jaxfg.core.FactorBase] = [
+        jaxfg.geometry.PriorFactor.make(
+            variable=pose_variables[0],
+            mu=constrained_to[0],
+            noise_model=jaxfg.noises.DiagonalGaussian.make_from_covariance(
+                diagonal=1e-40 * jnp.ones(3)
```
Comment (SuperN1ck) [inline tests/test_integration.py:224]: This would cause trouble in the current implementation

File: jaxfg/noises/_gaussians.py (hunk 1 above)
Comment (SuperN1ck) [reply jaxfg/noises/_gaussians.py]: Yes, totally agree, having `None` as default is nonsense. 

Thank you for the pointers :+1:

**Comment by brentyi** [comment]
Closing as per conversation -- maybe revisit later!
